src/VisualStudio/Core/Def/Diagnostics/RedirectingAnalyzerAssemblyResolver.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.IO;
+using System.Reflection;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics;
+
+[Export(typeof(IAnalyzerAssemblyResolver)), Shared]
+internal sealed class RedirectingAnalyzerAssemblyResolver : IAnalyzerAssemblyResolver
+{
+    private readonly VisualStudioWorkspaceImpl _workspace;
+    private readonly Lazy<ImmutableArray<Matcher>> _matchers;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public RedirectingAnalyzerAssemblyResolver(VisualStudioWorkspaceImpl workspace)
+    {
+        _workspace = workspace;
+        _matchers = new(CreateMatchers);
+    }
+
+    public Assembly? ResolveAssembly(AssemblyName assemblyName, string assemblyOriginalDirectory)
+    {
+        foreach (var matcher in _matchers.Value)
+        {
+            if (matcher.TryRedirect(assemblyOriginalDirectory) is { } redirectedDirectory)
+            {
+                var redirectedPath = Path.Combine(matcher.TargetDirectory, redirectedDirectory, assemblyName.Name + ".dll");
+                return Assembly.LoadFile(redirectedPath);
+            }
+        }
+
+        return null;
+    }
+
+    private ImmutableArray<Matcher> CreateMatchers()
```

Comment:
This could be improved (e.g., by creating regexes), but I'm not sure the spec is final yet.